pkg/oauth2/token_source.go

```diff
@@ -65,26 +70,48 @@ func (c *passwordCredentialsTokenSource) token(now func() time.Time) (*oauth2.To
 		if tok.RefreshToken == c.refreshToken.RefreshToken {
 			return tok, nil
 		}
-	} else {
-		tok, err = c.cfg.PasswordCredentialsToken(c.ctx, c.username, c.password)
+
+		err = c.setRefreshToken(tok)
 		if err != nil {
-			return nil, fmt.Errorf("password credentials token source failed: %v", err)
+			return nil, err
 		}
 
-		c.accessTokenSource = c.cfg.TokenSource(c.ctx, tok)
+		return tok, nil
+	}
+
+	return c.passwordCredentialsToken()
+}
+
+func (c *passwordCredentialsTokenSource) passwordCredentialsToken() (*oauth2.Token, error) {
+	c.grantsCounter()
```

Comment:
it seems to me we may want to differentiate in a successful and errored token retrieval no?

Reply:
We have that indication (status code returned by tollbooth) already in the `oauth` client metric, i.e. `client_api_requests_total{client="oauth",code="..."}`. This metric is also used to set up alerting.

Would it be idiomatic to add a `err="false|true"` here?

Reply:
but it's differentiated in terms of refresh token use and credentials flow? if so lgtm

Reply:
As discussed oob:
1. The existing `client_api_requests_total` metric does not differentiate the cause of the refresh. I added the `cause` label for that reason.
2. We'll add a `status` label here to have more insight on the result of the password grant operation.

ptal